<a href="https://colab.research.google.com/github/JuanSamaniego02/curso-ia-para-economia/blob/main/Taller_Limpieza_Alan_Rios_JuanSamaniego.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Taller 5 Manipulación y transformación de datos


Ejercicio 1

1. Importe todas las librerías que vaya a utilizar.

In [1]:
import os
import numpy as np
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

2. Mejore la visualización de los dataframes como se ha visto en clase.

In [2]:

pd.options.display.max_columns = None
pd.options.display.float_format = '{:,.2f}'.format

3. Conéctese a su Google Drive y establezca la ruta del dataset.

In [3]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [4]:
ruta = "/content/drive/MyDrive/Datasets IA"

4. Cargue los datos en el notebook. Llame al dataframe 'df_geih'

In [5]:
df_geih = pd.read_csv(ruta + "/GEIH_2025_Jul_Dic_Consolidado.csv", encoding = 'latin-1')

5. Obtenga las dimensiones de la base

In [6]:
df_geih.shape

(409127, 314)

In [8]:
df_geih.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 271065 entries, 0 to 271064
Columns: 314 entries, PERIODO to NOMBRE_DPTO
dtypes: float64(247), int64(46), object(21)
memory usage: 649.4+ MB


# Ejercicio 2

1. Crea un nuevo DataFrame llamado df_empleo que contenga únicamente las columnas que están en la lista columnas_interes.

In [7]:
columnas_interes = ['DIRECTORIO', 'SECUENCIA_P', 'ORDEN', 'MES', 'CLASE', 'DPTO', 'FEX_C18', 'PT', 'P3271', 'P6040', 'P6080', 'P3043', 'OCI',
                    'P6430', 'RAMA4D_R4', 'INGLABO', 'DESCRIPCION_CIIU', 'NOMBRE_DPTO']

In [8]:
df_empleo = df_geih[columnas_interes]

In [9]:
df_empleo

,DIRECTORIO,SECUENCIA_P,ORDEN,MES,CLASE,DPTO,FEX_C18,PT,P3271,P6040,P6080,P3043,OCI,P6430,RAMA4D_R4,INGLABO,DESCRIPCION_CIIU,NOMBRE_DPTO
0,8283143,1,1,7,2,8,27.75,1,1,72,6,NaN,1.00,7.00,143.00,"790,000.00",Cría de ovejas y cabras,ATLÁNTICO
1,8283143,1,2,7,2,8,27.75,1,2,34,6,NaN,NaN,NaN,NaN,NaN,NaN,ATLÁNTICO
2,8283145,1,1,7,2,8,59.11,1,1,45,6,NaN,1.00,1.00,141.00,"1,100,000.00",Cría de ganado bovino y bufalino,ATLÁNTICO
3,8283145,1,2,7,2,8,59.11,1,2,50,6,NaN,NaN,NaN,NaN,NaN,NaN,ATLÁNTICO
4,8283145,1,3,7,2,8,59.11,1,1,19,6,NaN,1.00,7.00,141.00,"820,000.00",Cría de ganado bovino y bufalino,ATLÁNTICO
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
409122,8473907,1,1,12,2,73,378.31,1,2,61,6,NaN,1.00,3.00,"9,700.00","1,400,000.00",Actividades de los hogares individuales como e...,TOLIMA
409123,8473908,1,1,12,1,54,534.55,1,1,58,6,9.00,1.00,2.00,"8,530.00","3,000,000.00",Establecimientos que combinan diferentes nivel...,NORTE DE SANTANDER
409124,8473908,1,2,12,1,54,534.55,1,2,51,6,9.00,1.00,2.00,"8,412.00","7,956,000.00",Actividades ejecutivas de la administración pú...,NORTE DE SANTANDER
409125,8473908,1,3,12,1,54,534.55,1,2,31,6,7.00,NaN,NaN,NaN,NaN,NaN,NORTE DE SANTANDER


2. Renombra las columnas del DataFrame df_empleo utilizando el diccionario mapa_nombres.

In [10]:
mapa_nombres = {
'DIRECTORIO': 'id_vivienda',
'SECUENCIA_P': 'id_hogar',
'ORDEN': 'id_persona',
'MES': 'mes',
'CLASE': 'clase',
'DPTO': 'departamento',
'FEX_C18': 'factor_expansion',
'PT': 'poblacion_total',
'P3271': 'sexo',
'P6040': 'edad',
'P6080': 'grupo_etnico',
'P3043': 'nivel_educativo',
'OCI': 'ocupados',
'P6430': 'categoria_ocupacional',
'RAMA4D_R4': 'cod_ciiu',
'INGLABO': 'ingreso_laboral',
'DESCRIPCION_CIIU': 'desc_ciiu',
'NOMBRE_DPTO': 'nombre_dpto'}

In [11]:
df_empleo = df_empleo.rename(columns=mapa_nombres)

3. Obtenga las dimensiones de la base

In [12]:
df_empleo.shape

(409127, 18)

In [13]:
df_empleo.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 409127 entries, 0 to 409126
Data columns (total 18 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   id_vivienda            409127 non-null  int64  
 1   id_hogar               409127 non-null  int64  
 2   id_persona             409127 non-null  int64  
 3   mes                    409127 non-null  int64  
 4   clase                  409127 non-null  int64  
 5   departamento           409127 non-null  int64  
 6   factor_expansion       409127 non-null  float64
 7   poblacion_total        409127 non-null  int64  
 8   sexo                   409127 non-null  int64  
 9   edad                   409127 non-null  int64  
 10  grupo_etnico           409127 non-null  int64  
 11  nivel_educativo        208679 non-null  float64
 12  ocupados               181853 non-null  float64
 13  categoria_ocupacional  181853 non-null  float64
 14  cod_ciiu               181853 non-nu

In [14]:
df_geih.shape

(409127, 314)

In [15]:
df_geih.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 409127 entries, 0 to 409126
Columns: 314 entries, PERIODO to NOMBRE_DPTO
dtypes: float64(247), int64(46), object(21)
memory usage: 980.1+ MB


# Ejercicio 3

1. Verifica si hay filas duplicadas en df_empleo.

In [16]:
df_empleo.duplicated()

,0
0,False
1,False
2,False
3,False
4,False
...,...
409122,False
409123,False
409124,False
409125,False


2. Cuenta cuántas filas duplicadas existen.

In [17]:
df_empleo.duplicated().sum()

np.int64(0)

3. Elimina las filas duplicadas, si las hay.

In [18]:
# No hay filas duplicadas

4. Verifica si hay filas duplicadas específicamente en las columnas 'id_vivienda', 'id_hogar', 'id_persona'.

In [19]:
df_empleo.duplicated(subset=['id_vivienda', 'id_hogar', 'id_persona'])

,0
0,False
1,False
2,False
3,False
4,False
...,...
409122,False
409123,False
409124,False
409125,False


In [20]:
df_empleo.duplicated(subset=['id_vivienda', 'id_hogar', 'id_persona']).sum()

np.int64(0)

# Ejercicio 4

1. Calcula el número de valores nulos (NaN) en la columna ingreso_laboral antes de cualquier modificación.

In [21]:
df_empleo["ingreso_laboral"].isna().sum()

np.int64(231144)

2. En la GEIH, un ingreso faltante a menudo significa que la persona no reportó ingresos o no aplica (ej. no trabaja). Una estrategia común es imputar estos valores con 0. Rellena los valores NaN en la columna ingreso_laboral con 0.

In [22]:
df_empleo["ingreso_laboral"].fillna(0, inplace = True)

# Ejercicio 5

Las variables categóricas a menudo necesitan ser "traducidas" para hacerlas más entendibles para el análisis.

1. Usa los siguientes diccionarios para reemplazar los códigos numéricos por etiquetas de texto en las siguientes columnas: 'clase', 'sexo', 'grupo_etnico', 'nivel_educativo' y 'categoria_ocupacional'. Muestra los conteos de las nuevas categorías usando .value_counts() para verificar el cambio en cada una de las columnas.

In [23]:
dic_clase = {1: 'Cabeceras', 2: 'Centros poblados y rural disperso'}
dic_sexo = {1: 'Hombre', 2: 'Mujer'}
dic_etnico = {1: 'Indígena', 2: 'Gitano (Rom)', 3: 'Raizal del archipiélago', 4: 'Palenquero',
              5: 'Negro, mulato (afrodescendiente)', 6: 'Ninguno de los anteriores'}
dic_nivel_educativo = {1: 'Ninguno', 2: 'Educación media', 3: 'Educación media',
                       4: 'Educación media', 5: 'Técnica profesional', 6: 'Tecnológica',
                       7: 'Universitaria', 8: 'Especialización', 9: 'Maestría',
                       10: 'Doctorado', 99: 'No sabe, no informa'}
dic_cat_ocu = {1: 'Asalariado', 2: 'Asalariado', 3: 'Asalariado',
               8: 'Asalariado', 4: 'Independiente', 5: 'Independiente',
               9: 'Independiente', 6: 'Trabajador sin remuneración',
               7: 'Trabajador sin remuneración'}

In [24]:
df_empleo["clase"] = df_empleo["clase"].replace(dic_clase)
df_empleo["sexo"] = df_empleo["sexo"].replace(dic_sexo)
df_empleo["grupo_etnico"] = df_empleo["grupo_etnico"].replace(dic_etnico)
df_empleo["nivel_educativo"] = df_empleo["nivel_educativo"].replace(dic_nivel_educativo)
df_empleo["categoria_ocupacional"] = df_empleo["categoria_ocupacional"].replace(dic_cat_ocu)

In [25]:
print(df_empleo["clase"].value_counts())


clase
Cabeceras                            356377
Centros poblados y rural disperso     52750
Name: count, dtype: int64


In [26]:
print(df_empleo["sexo"].value_counts())

sexo
Mujer     218597
Hombre    190530
Name: count, dtype: int64


In [27]:
print(df_empleo["grupo_etnico"].value_counts())

grupo_etnico
Ninguno de los anteriores           350562
Negro, mulato (afrodescendiente)     39032
Indígena                             18420
Raizal del archipiélago                925
Palenquero                             120
Gitano (Rom)                            68
Name: count, dtype: int64


In [28]:
print(df_empleo["nivel_educativo"].value_counts())

nivel_educativo
Educación media        116635
Universitaria           32517
Técnica profesional     24503
Ninguno                 12963
Tecnológica             10640
Especialización          7927
Maestría                 3154
Doctorado                 339
No sabe, no informa         1
Name: count, dtype: int64


In [29]:
print(df_empleo["categoria_ocupacional"].value_counts())

categoria_ocupacional
Asalariado                     92016
Independiente                  82498
Trabajador sin remuneración     7339
Name: count, dtype: int64


Ejercicio 6

1.  Reorganiza las columnas con el siguiente orden: 'id_vivienda', 'id_hogar', 'id_persona', 'mes', 'clase', 'departamento', 'nombre_dpto', 'factor_expansion', 'poblacion_total', 'sexo', 'edad', 'grupo_etnico', 'nivel_educativo', 'ocupados', 'categoria_ocupacional', 'cod_ciiu', 'desc_ciiu', 'ingreso_laboral'

In [30]:
orden_columnas = ['id_vivienda', 'id_hogar', 'id_persona', 'mes', 'clase','departamento', 'nombre_dpto', 'factor_expansion', 'poblacion_total',
'sexo', 'edad', 'grupo_etnico', 'nivel_educativo', 'ocupados','categoria_ocupacional', 'cod_ciiu', 'desc_ciiu', 'ingreso_laboral']



In [31]:
df_empleo = df_empleo[orden_columnas]

In [32]:
df_empleo.columns

Index(['id_vivienda', 'id_hogar', 'id_persona', 'mes', 'clase', 'departamento',
       'nombre_dpto', 'factor_expansion', 'poblacion_total', 'sexo', 'edad',
       'grupo_etnico', 'nivel_educativo', 'ocupados', 'categoria_ocupacional',
       'cod_ciiu', 'desc_ciiu', 'ingreso_laboral'],
      dtype='object')

# Ejercicio 7

Para realizar cálculos y análisis correctamente, las columnas deben tener el tipo de dato adecuado (numérico, categórico, etc.).

1. Muestra en pantalla los tipos de datos actuales de df_empleo.

In [33]:
df_empleo.dtypes

,0
id_vivienda,int64
id_hogar,int64
id_persona,int64
mes,int64
clase,object
departamento,int64
nombre_dpto,object
factor_expansion,float64
poblacion_total,int64
sexo,object


2. Las siguientes columnas representan categorías: 'clase', 'departamento', 'nombre_dpto', 'sexo', 'grupo_etnico', 'nivel_educativo', 'categoria_ocupacional', 'cod_ciiu', 'desc_ciiu'. Conviértelas al tipo category para optimizar el uso de memoria y facilitar análisis futuros.

In [34]:
columnas_categoricas = ['clase', 'departamento', 'nombre_dpto', 'sexo', 'grupo_etnico', 'nivel_educativo', 'categoria_ocupacional', 'cod_ciiu', 'desc_ciiu']

In [35]:
for col in columnas_categoricas:
    df_empleo[col] = df_empleo[col].astype('category')

In [36]:
df_empleo.dtypes

,0
id_vivienda,int64
id_hogar,int64
id_persona,int64
mes,int64
clase,category
departamento,category
nombre_dpto,category
factor_expansion,float64
poblacion_total,int64
sexo,category


# Ejercicio 8

1. Ordena el dataframe por la columna "edad" de mayor a menor. Observa la columna 'edad' y escribe en una celda de markdown cuál fue la edad mínima y la edad máxima de las personas que respondieron la encuesta de la GEIH.

In [37]:
df_empleo = df_empleo.sort_values(by="edad", ascending=False)

In [38]:
edad_min = df_empleo["edad"].min()
edad_max = df_empleo["edad"].max()

print("Edad mínima:", edad_min)
print("Edad máxima:", edad_max)

Edad mínima: 0
Edad máxima: 108


La edad mínima de las personas en la encuesta es de 0 años, lo que indica que puede tratarse de bebés, ya que la GEIH dentro de su recolección incluye a toda la población, mientras que la edad máxima es de 108 años.

# Ejercicio 9

Agrupar los datos por categorías, nos permite realizar análisis más profundos y obtener insights valiosos, que a simple vista no vemos en la base de datos.

1.Crea un nuevo dataframe llamado "df_sexo" donde se agrupe por 'sexo' y se calcule la suma de las columnas 'poblacion_total' y 'ocupados'. Crea una nueva columna que se llame 'porcentaje_ocupados' donde dividas las columna 'ocupados' sobre 'poblacion_total'. Al final usa reset_index() para convertir las categorías de sexo en una columna.

In [39]:
df_sexo = df_empleo.groupby("sexo")[["poblacion_total", "ocupados"]].sum()
df_sexo["porcentaje_ocupados"] = df_sexo["ocupados"] / df_sexo["poblacion_total"]
df_sexo = df_sexo.reset_index()

In [40]:
df_sexo

,sexo,poblacion_total,ocupados,porcentaje_ocupados
0,Hombre,190530,"100,790.00",0.53
1,Mujer,218597,"81,063.00",0.37


# Ejercicio 10

1. A partir de df_empleo, crea el DataFrame df_ing con observaciones donde ingreso_laboral no sea nulo y además sea mayor estrictamente al salario mínimo para 2026 (con subsidio de transporte).

In [41]:
salario_minimo_2026 = 2000000
df_ing = df_empleo[(df_empleo["ingreso_laboral"].notnull()) & (df_empleo["ingreso_laboral"] > salario_minimo_2026)]

In [42]:
df_ing.shape
df_ing["ingreso_laboral"].min()

2003000.0

2. ¿Cuántas personas tienen un salario estrictamente mayor al salario mínimo?

In [43]:
len(df_ing)

38093

3. Crea el dataframe df_mujeres donde se encuentren solamente mujeres que están ubicadas en el departamento del Tolima, que son asalariadas y cuyo nivel educativo es educación media o superior.

In [44]:
df_empleo["sexo"] = df_empleo["sexo"].str.strip().str.lower()
df_empleo["nombre_dpto"] = df_empleo["nombre_dpto"].str.strip().str.lower()
df_empleo["categoria_ocupacional"] = df_empleo["categoria_ocupacional"].str.strip().str.lower()
df_empleo["nivel_educativo"] = df_empleo["nivel_educativo"].str.strip()

df_mujeres = df_empleo[
    (df_empleo["sexo"] == "mujer") &
    (df_empleo["nombre_dpto"] == "tolima") &
    (df_empleo["categoria_ocupacional"] == "asalariado") &
    (df_empleo["nivel_educativo"].isin([
        "Educación media",
        "Técnica profesional",
        "Tecnológica",
        "Universitaria",
        "Especialización",
        "Maestría",
        "Doctorado"
    ]))
]
df_mujeres.head()


,id_vivienda,id_hogar,id_persona,mes,clase,departamento,nombre_dpto,factor_expansion,poblacion_total,sexo,edad,grupo_etnico,nivel_educativo,ocupados,categoria_ocupacional,cod_ciiu,desc_ciiu,ingreso_laboral
204009,8373281,1,1,9,Cabeceras,73,tolima,309.84,1,mujer,76,Ninguno de los anteriores,Educación media,1.00,asalariado,"8,530.00",Establecimientos que combinan diferentes nivel...,"1,423,500.00"
341096,8435212,1,1,11,Cabeceras,73,tolima,379.47,1,mujer,68,Ninguno de los anteriores,Técnica profesional,1.00,asalariado,"6,820.00",Actividades inmobiliarias realizadas a cambio ...,"950,000.00"
83927,8323356,1,1,8,Cabeceras,73,tolima,407.65,1,mujer,67,Ninguno de los anteriores,Especialización,1.00,asalariado,"8,530.00",Establecimientos que combinan diferentes nivel...,"5,700,000.00"
391232,8465937,1,2,12,Cabeceras,73,tolima,310.36,1,mujer,67,Ninguno de los anteriores,Especialización,1.00,asalariado,"8,530.00",Establecimientos que combinan diferentes nivel...,"4,700,000.00"
226543,8383161,1,1,10,Cabeceras,73,tolima,285.20,1,mujer,67,Ninguno de los anteriores,Universitaria,1.00,asalariado,"8,530.00",Establecimientos que combinan diferentes nivel...,"3,500,000.00"


In [45]:
df_empleo["sexo"] == "Mujer"

,sexo
256896,False
5135,False
32904,False
256662,False
396633,False
...,...
245532,False
145507,False
185317,False
366733,False


In [46]:
df_empleo["nombre_dpto"] == "Tolima"

,nombre_dpto
256896,False
5135,False
32904,False
256662,False
396633,False
...,...
245532,False
145507,False
185317,False
366733,False


In [47]:
df_empleo["categoria_ocupacional"] == "Asalariado"

,categoria_ocupacional
256896,False
5135,False
32904,False
256662,False
396633,False
...,...
245532,False
145507,False
185317,False
366733,False


In [48]:
df_empleo["nivel_educativo"].isin(["Educación media", "Técnica profesional", "Tecnológica", "Universitaria", "Especialización", "Maestría", "Doctorado"])

,nivel_educativo
256896,False
5135,False
32904,False
256662,False
396633,False
...,...
245532,False
145507,False
185317,False
366733,False


In [49]:
df_mujeres.shape
df_mujeres.head()

,id_vivienda,id_hogar,id_persona,mes,clase,departamento,nombre_dpto,factor_expansion,poblacion_total,sexo,edad,grupo_etnico,nivel_educativo,ocupados,categoria_ocupacional,cod_ciiu,desc_ciiu,ingreso_laboral
204009,8373281,1,1,9,Cabeceras,73,tolima,309.84,1,mujer,76,Ninguno de los anteriores,Educación media,1.00,asalariado,"8,530.00",Establecimientos que combinan diferentes nivel...,"1,423,500.00"
341096,8435212,1,1,11,Cabeceras,73,tolima,379.47,1,mujer,68,Ninguno de los anteriores,Técnica profesional,1.00,asalariado,"6,820.00",Actividades inmobiliarias realizadas a cambio ...,"950,000.00"
83927,8323356,1,1,8,Cabeceras,73,tolima,407.65,1,mujer,67,Ninguno de los anteriores,Especialización,1.00,asalariado,"8,530.00",Establecimientos que combinan diferentes nivel...,"5,700,000.00"
391232,8465937,1,2,12,Cabeceras,73,tolima,310.36,1,mujer,67,Ninguno de los anteriores,Especialización,1.00,asalariado,"8,530.00",Establecimientos que combinan diferentes nivel...,"4,700,000.00"
226543,8383161,1,1,10,Cabeceras,73,tolima,285.20,1,mujer,67,Ninguno de los anteriores,Universitaria,1.00,asalariado,"8,530.00",Establecimientos que combinan diferentes nivel...,"3,500,000.00"


4. ¿Cuáles son las dimensiones de df_mujeres?

In [50]:
df_mujeres.shape

(1137, 18)

In [51]:
df_mujeres.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1137 entries, 204009 to 384668
Data columns (total 18 columns):
 #   Column                 Non-Null Count  Dtype   
---  ------                 --------------  -----   
 0   id_vivienda            1137 non-null   int64   
 1   id_hogar               1137 non-null   int64   
 2   id_persona             1137 non-null   int64   
 3   mes                    1137 non-null   int64   
 4   clase                  1137 non-null   category
 5   departamento           1137 non-null   category
 6   nombre_dpto            1137 non-null   object  
 7   factor_expansion       1137 non-null   float64 
 8   poblacion_total        1137 non-null   int64   
 9   sexo                   1137 non-null   object  
 10  edad                   1137 non-null   int64   
 11  grupo_etnico           1137 non-null   category
 12  nivel_educativo        1137 non-null   object  
 13  ocupados               1137 non-null   float64 
 14  categoria_ocupacional  1137 non-null  

In [52]:
df_mujeres.head()

,id_vivienda,id_hogar,id_persona,mes,clase,departamento,nombre_dpto,factor_expansion,poblacion_total,sexo,edad,grupo_etnico,nivel_educativo,ocupados,categoria_ocupacional,cod_ciiu,desc_ciiu,ingreso_laboral
204009,8373281,1,1,9,Cabeceras,73,tolima,309.84,1,mujer,76,Ninguno de los anteriores,Educación media,1.00,asalariado,"8,530.00",Establecimientos que combinan diferentes nivel...,"1,423,500.00"
341096,8435212,1,1,11,Cabeceras,73,tolima,379.47,1,mujer,68,Ninguno de los anteriores,Técnica profesional,1.00,asalariado,"6,820.00",Actividades inmobiliarias realizadas a cambio ...,"950,000.00"
83927,8323356,1,1,8,Cabeceras,73,tolima,407.65,1,mujer,67,Ninguno de los anteriores,Especialización,1.00,asalariado,"8,530.00",Establecimientos que combinan diferentes nivel...,"5,700,000.00"
391232,8465937,1,2,12,Cabeceras,73,tolima,310.36,1,mujer,67,Ninguno de los anteriores,Especialización,1.00,asalariado,"8,530.00",Establecimientos que combinan diferentes nivel...,"4,700,000.00"
226543,8383161,1,1,10,Cabeceras,73,tolima,285.20,1,mujer,67,Ninguno de los anteriores,Universitaria,1.00,asalariado,"8,530.00",Establecimientos que combinan diferentes nivel...,"3,500,000.00"
